File: nanobot/nanobot/interagent/performance/pool.py
"""
对象池与Buffer池
用于内存优化，减少GC压力

@module interagent.performance.pool
@version 1.0.0
"""

from dataclasses import dataclass, field
from typing import Any, Callable, Generic, List, Optional, TypeVar

T = TypeVar("T")


# ============================================================================
# Types
# ============================================================================


@dataclass
class PoolStats:
    """池统计信息"""

    available: int
    """池中可用对象数"""

    in_use: int
    """正在使用的对象数"""

    total_created: int
    """总创建数"""

    total_acquired: int
    """总获取数"""

    total_released: int
    """总释放数"""

    hit_rate: float
    """命中率"""


# ============================================================================
# ObjectPool
# ============================================================================


class ObjectPool(Generic[T]):
    """
    通用对象池

    Example:
        pool = ObjectPool(
            factory=lambda: {"data": "", "count": 0},
            reset=lambda obj: obj.update({"data": "", "count": 0}),
            max_size=100
        )

        obj = pool.acquire()
        # 使用对象...
        pool.release(obj)
    """

    def __init__(
        self,
        factory: Callable[[], T],
        reset: Optional[Callable[[T], None]] = None,
        initial_size: int = 10,
        max_size: int = 100,
    ):
        """
        初始化对象池

        Args:
            factory: 工厂函数 - 创建新对象
            reset: 重置函数 - 重置对象状态
            initial_size: 初始池大小
            max_size: 最大池大小
        """
        self._factory = factory
        self._reset = reset
        self._max_size = max_size

        self._pool: List[T] = []
        self._in_use_count = 0
        self._total_created = 0
        self._total_acquired = 0
        self._total_released = 0
        self._hits = 0
        self._misses = 0

        # 预创建对象
        for _ in range(initial_size):
            self._pool.append(self._factory())
            self._total_created += 1

    def acquire(self) -> T:
        """获取对象"""
        self._total_acquired += 1

        if self._pool:
            self._hits += 1
            obj = self._pool.pop()
            self._in_use_count += 1
            return obj

        self._misses += 1
        self._total_created += 1
        self._in_use_count += 1
        return self._factory()

    def release(self, obj: T) -> None:
        """释放对象回池"""
        if self._in_use_count <= 0:
            return

        self._total_released += 1
        self._in_use_count -= 1

        if len(self._pool) < self._max_size:
            if self._reset:
                self._reset(obj)
            self._pool.append(obj)
        # 超过最大容量则丢弃

    def acquire_batch(self, count: int) -> List[T]:
        """批量获取对象"""
        return [self.acquire() for _ in range(count)]

    def release_batch(self, objects: List[T]) -> None:
        """批量释放对象"""
        for obj in objects:
            self.release(obj)

    def clear(self) -> None:
        """清空池"""
        self._pool = []
        self._in_use_count = 0

    def get_stats(self) -> PoolStats:
        """获取统计信息"""
        total_requests = self._hits + self._misses
        return PoolStats(
            available=len(self._pool),
            in_use=self._in_use_count,
            total_created=self._total_created,
            total_acquired=self._total_acquired,
            total_released=self._total_released,
            hit_rate=self._hits / total_requests if total_requests > 0 else 0,
        )

    @property
    def available(self) -> int:
        """获取可用对象数"""
        return len(self._pool)

    @property
    def in_use(self) -> int:
        """获取使用中对象数"""
        return self._in_use_count


# ============================================================================
# BufferPool
# ============================================================================


class BufferPool:
    """
    Buffer 池
    专门用于 bytes 复用

    Example:
        buffer_pool = BufferPool(buffer_size=4096, max_size=50)

        buffer = buffer_pool.acquire()
        # 使用 buffer...
        buffer_pool.release(buffer)
    """

    def __init__(
        self,
        buffer_size: int = 4096,
        initial_size: int = 10,
        max_size: int = 50,
    ):
        """
        初始化 Buffer 池

        Args:
            buffer_size: Buffer 大小
            initial_size: 初始池大小
            max_size: 最大池大小
        """
        self._buffer_size = buffer_size
        self._max_size = max_size

        self._pool: List[bytearray] = []
        self._in_use_count = 0
        self._total_created = 0
        self._total_acquired = 0
        self._total_released = 0
        self._hits = 0
        self._misses = 0

        # 预创建 buffers
        for _ in range(initial_size):
            self._pool.append(bytearray(buffer_size))
            self._total_created += 1

    def acquire(self) -> bytearray:
        """获取 Buffer"""
        self._total_acquired += 1

        if self._pool:
            self._hits += 1
            buffer = self._pool.pop()
            self._in_use_count += 1
            # 清空 buffer
            for i in range(len(buffer)):
                buffer[i] = 0
            return buffer

        self._misses += 1
        self._total_created += 1
        self._in_use_count += 1
        return bytearray(self._buffer_size)

    def release(self, buffer: bytearray) -> None:
        """释放 Buffer 回池"""
        if self._in_use_count <= 0:
            return

        self._total_released += 1
        self._in_use_count -= 1

        if len(buffer) == self._buffer_size and len(self._pool) < self._max_size:
            self._pool.append(buffer)

    def clear(self) -> None:
        """清空池"""
        self._pool = []
        self._in_use_count = 0

    def get_stats(self) -> PoolStats:
        """获取统计信息"""
        total_requests = self._hits + self._misses
        return PoolStats(
            available=len(self._pool),
            in_use=self._in_use_count,
            total_created=self._total_created,
            total_acquired=self._total_acquired,
            total_released=self._total_released,
            hit_rate=self._hits / total_requests if total_requests > 0 else 0,
        )

    @property
    def buffer_size(self) -> int:
        """获取 Buffer 大小"""
        return self._buffer_size

    @property
    def available(self) -> int:
        """获取可用 Buffer 数"""
        return len(self._pool)

    @property
    def in_use(self) -> int:
        """获取使用中 Buffer 数"""
        return self._in_use_count


# ============================================================================
# Helper Functions
# ============================================================================


def create_object_pool(
    factory: Callable[[], T],
    reset: Optional[Callable[[T], None]] = None,
    initial_size: int = 10,
    max_size: int = 100,
) -> ObjectPool[T]:
    """创建对象池"""
    return ObjectPool(
        factory=factory,
        reset=reset,
        initial_size=initial_size,
        max_size=max_size,
    )


def create_buffer_pool(
    buffer_size: int = 4096,
    initial_size: int = 10,
    max_size: int = 50,
) -> BufferPool:
    """创建 Buffer 池"""
    return BufferPool(
        buffer_size=buffer_size,
        initial_size=initial_size,
        max_size=max_size,
    )

File: automaton/src/interagent/performance/pool.ts
/**
 * 对象池与Buffer池
 * 用于内存优化，减少GC压力
 *
 * @module interagent.performance.pool
 * @version 1.0.0
 */

// ============================================================================
// Types
// ============================================================================

export interface PoolOptions<T> {
  /** 初始池大小 */
  initialSize?: number;
  /** 最大池大小 */
  maxSize?: number;
  /** 工厂函数 - 创建新对象 */
  factory: () => T;
  /** 重置函数 - 重置对象状态 */
  reset?: (obj: T) => void;
}

export interface BufferPoolOptions {
  /** Buffer 大小 */
  bufferSize: number;
  /** 初始池大小 */
  initialSize?: number;
  /** 最大池大小 */
  maxSize?: number;
}

export interface PoolStats {
  /** 池中可用对象数 */
  available: number;
  /** 正在使用的对象数 */
  inUse: number;
  /** 总创建数 */
  totalCreated: number;
  /** 总获取数 */
  totalAcquired: number;
  /** 总释放数 */
  totalReleased: number;
  /** 命中率 */
  hitRate: number;
}

// ============================================================================
// ObjectPool
// ============================================================================

/**
 * 通用对象池
 *
 * Example:
 * ```typescript
 * const pool = new ObjectPool({
 *   factory: () => ({ data: '', count: 0 }),
 *   reset: (obj) => { obj.data = ''; obj.count = 0; },
 *   maxSize: 100
 * });
 *
 * const obj = pool.acquire();
 * // 使用对象...
 * pool.release(obj);
 * ```
 */
export class ObjectPool<T> {
  private pool: T[] = [];
  private inUseCount = 0;
  private totalCreated = 0;
  private totalAcquired = 0;
  private totalReleased = 0;
  private hits = 0;
  private misses = 0;

  private readonly factory: () => T;
  private readonly reset?: (obj: T) => void;
  private readonly maxSize: number;

  constructor(options: PoolOptions<T>) {
    this.factory = options.factory;
    this.reset = options.reset;
    this.maxSize = options.maxSize ?? 100;

    const initialSize = options.initialSize ?? 10;
    for (let i = 0; i < initialSize; i++) {
      this.pool.push(this.factory());
      this.totalCreated++;
    }
  }

  /**
   * 获取对象
   */
  acquire(): T {
    this.totalAcquired++;

    if (this.pool.length > 0) {
      this.hits++;
      const obj = this.pool.pop()!;
      this.inUseCount++;
      return obj;
    }

    this.misses++;
    this.totalCreated++;
    this.inUseCount++;
    return this.factory();
  }

  /**
   * 释放对象回池
   */
  release(obj: T): void {
    if (this.inUseCount <= 0) {
      return;
    }

    this.totalReleased++;
    this.inUseCount--;

    if (this.pool.length < this.maxSize) {
      if (this.reset) {
        this.reset(obj);
      }
      this.pool.push(obj);
    }
    // 超过最大容量则丢弃
  }

  /**
   * 批量获取对象
   */
  acquireBatch(count: number): T[] {
    const result: T[] = [];
    for (let i = 0; i < count; i++) {
      result.push(this.acquire());
    }
    return result;
  }

  /**
   * 批量释放对象
   */
  releaseBatch(objects: T[]): void {
    for (const obj of objects) {
      this.release(obj);
    }
  }

  /**
   * 清空池
   */
  clear(): void {
    this.pool = [];
    this.inUseCount = 0;
  }

  /**
   * 获取统计信息
   */
  getStats(): PoolStats {
    const totalRequests = this.hits + this.misses;
    return {
      available: this.pool.length,
      inUse: this.inUseCount,
      totalCreated: this.totalCreated,
      totalAcquired: this.totalAcquired,
      totalReleased: this.totalReleased,
      hitRate: totalRequests > 0 ? this.hits / totalRequests : 0,
    };
  }

  /**
   * 获取可用对象数
   */
  get available(): number {
    return this.pool.length;
  }

  /**
   * 获取使用中对象数
   */
  get inUse(): number {
    return this.inUseCount;
  }
}

// ============================================================================
// BufferPool
// ============================================================================

/**
 * Buffer 池
 * 专门用于 Buffer 复用
 *
 * Example:
 * ```typescript
 * const bufferPool = new BufferPool({ bufferSize: 4096, maxSize: 50 });
 *
 * const buffer = bufferPool.acquire();
 * // 使用 buffer...
 * bufferPool.release(buffer);
 * ```
 */
export class BufferPool {
  private pool: Buffer[] = [];
  private inUseCount = 0;
  private totalCreated = 0;
  private totalAcquired = 0;
  private totalReleased = 0;
  private hits = 0;
  private misses = 0;

  private readonly bufferSize: number;
  private readonly maxSize: number;

  constructor(options: BufferPoolOptions) {
    this.bufferSize = options.bufferSize;
    this.maxSize = options.maxSize ?? 50;

    const initialSize = options.initialSize ?? 10;
    for (let i = 0; i < initialSize; i++) {
      this.pool.push(Buffer.allocUnsafe(this.bufferSize));
      this.totalCreated++;
    }
  }

  /**
   * 获取 Buffer
   */
  acquire(): Buffer {
    this.totalAcquired++;

    if (this.pool.length > 0) {
      this.hits++;
      const buffer = this.pool.pop()!;
      this.inUseCount++;
      buffer.fill(0); // 清空 buffer
      return buffer;
    }

    this.misses++;
    this.totalCreated++;
    this.inUseCount++;
    return Buffer.allocUnsafe(this.bufferSize);
  }

  /**
   * 释放 Buffer 回池
   */
  release(buffer: Buffer): void {
    if (this.inUseCount <= 0) {
      return;
    }

    this.totalReleased++;
    this.inUseCount--;

    if (buffer.length === this.bufferSize && this.pool.length < this.maxSize) {
      this.pool.push(buffer);
    }
  }

  /**
   * 清空池
   */
  clear(): void {
    this.pool = [];
    this.inUseCount = 0;
  }

  /**
   * 获取统计信息
   */
  getStats(): PoolStats {
    const totalRequests = this.hits + this.misses;
    return {
      available: this.pool.length,
      inUse: this.inUseCount,
      totalCreated: this.totalCreated,
      totalAcquired: this.totalAcquired,
      totalReleased: this.totalReleased,
      hitRate: totalRequests > 0 ? this.hits / totalRequests : 0,
    };
  }

  /**
   * 获取 Buffer 大小
   */
  getBufferSize(): number {
    return this.bufferSize;
  }

  /**
   * 获取可用 Buffer 数
   */
  get available(): number {
    return this.pool.length;
  }

  /**
   * 获取使用中 Buffer 数
   */
  get inUse(): number {
    return this.inUseCount;
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * 创建对象池
 */
export function createObjectPool<T>(options: PoolOptions<T>): ObjectPool<T> {
  return new ObjectPool<T>(options);
}

/**
 * 创建 Buffer 池
 */
export function createBufferPool(options: BufferPoolOptions): BufferPool {
  return new BufferPool(options);
}
